src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -292,15 +292,12 @@ private static SslClientAuthenticationOptions ConstructSslOptions(HttpConnection
         }
 
         private async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
-            WaitForAvailableHttp11Connection(TaskCompletionSource<HttpConnection> waiterTaskSource, HttpRequestMessage request, CancellationToken cancellationToken)
+            WaitForAvailableHttp11Connection(TaskCompletionSourceWithCancellation<HttpConnection> waiter, HttpRequestMessage request, CancellationToken cancellationToken)
         {
-            using (cancellationToken.Register(s => ((TaskCompletionSource<HttpConnection>)s).TrySetCanceled(), waiterTaskSource))
+            HttpConnection connection = await waiter.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
If it were me, I'd have manually inlined this method rather than creating a separate one, since it's only a few lines, it's not used elsewhere, and it avoids an extra layer of state machines.  But, up to you.